dll/shellext/netplwiz/SHDisconnectNetDrives.c

```diff
@@ -0,0 +1,206 @@
+    if (hShell32 == NULL)
+        return;
+    
+    hIconDrive = LoadImageW(hShell32, MAKEINTRESOURCEW(10), IMAGE_ICON,
+                            GetSystemMetrics(SM_CXSMICON),
+                            GetSystemMetrics(SM_CYSMICON), 0);
+    
+    if (hIconDrive == NULL)
+        return;
+    
+    hImgList = ImageList_Create(16, 16, ILC_COLOR32 | ILC_MASK, 1, 1);
+    
+    if (hImgList == NULL)
+    {
+        DestroyIcon(hIconDrive);
+        return;
+    }
+    
+    ImageList_AddIcon(hImgList, hIconDrive);
+    DestroyIcon(hIconDrive);
+    ListView_SetImageList(hListView, hImgList, LVSIL_SMALL);
+    
+    dRet = WNetOpenEnum(RESOURCE_CONNECTED, RESOURCETYPE_DISK, 0, NULL, &hEnum);
+    if (dRet != WN_SUCCESS)
+    {
+        return;
+    }
+
+    lpRes = HeapAlloc(GetProcessHeap(), 0, dwSize);
+    if (!lpRes)
+    {
+        WNetCloseEnum(hEnum);
+        return;
+    }
+
+    do
+    {
+        memset(lpRes, 0, dwSize);
+        dRet = WNetEnumResource(hEnum, &dwCount, lpRes, &dwSize);
+        if (dRet == WN_SUCCESS || dRet == WN_MORE_DATA)
+        {
+            lpCur = lpRes;
+            for (; dwCount; dwCount--)
+            {
+                if (!NULL || wcsicmp(lpCur->lpLocalName, NULL) == 0)
+                {
+                    memset(&item, 0, sizeof(item));
+                    item.mask = LVIF_TEXT | LVIF_IMAGE;
+                    item.iImage = 0;
+                    item.pszText = lpCur->lpLocalName;
+                    item.lParam = 0;
+                    item.iItem = ListView_InsertItem(hListView, &item);
+                    ListView_SetItemText(hListView, item.iItem, 1, lpCur->lpRemoteName);
+                }
+
+                lpCur++;
+            }
+        }
+    } while (dRet != WN_NO_MORE_ENTRIES);
+
+    HeapFree(GetProcessHeap(), 0, lpRes);
+    WNetCloseEnum(hEnum);
+}
+
+
+VOID UpdateButtonStatus(WPARAM wParam, LPARAM lParam, HWND hDlg)
+{
+    LPNMHDR pnmh;
+    HWND hListView = GetDlgItem(hDlg, IDC_CONNECTEDDRIVELIST);
+    HWND hOkbutton = GetDlgItem(hDlg, ID_OK);
+
+    pnmh = (LPNMHDR)lParam;
+
+    if (pnmh->hwndFrom == hListView)
+    {
+        switch (pnmh->code)
+        {
+            case LVN_ITEMCHANGED:
+                if (ListView_GetSelectedCount(hListView))
+                {
+                    EnableWindow(hOkbutton, TRUE);
+                }
+                else
+                {
+                    EnableWindow(hOkbutton, FALSE);
+                }
+                break;
+        }
+    }
+}
+
+DWORD DisconnectDriveExit(HWND hDlg)
+{
+    TCHAR driveLetter[10];
+    HRESULT hr;
+    HWND hListView = GetDlgItem(hDlg, IDC_CONNECTEDDRIVELIST);
+    INT nItem = ListView_GetNextItem(hListView, -1, LVNI_SELECTED);
+
+    if (nItem == -1)
+        return 0;
+
+    ListView_GetItemText(hListView, nItem, 0, driveLetter, ARRAYSIZE(driveLetter));
+    hr = WNetCancelConnection2(driveLetter, CONNECT_UPDATE_PROFILE, FALSE);
+
+    EndDialog(hDlg, ID_OK);
```

Comment:
Improvement suggestion (for later): display a message box in case the disconnection fails (and don't quit the dialog in that case?).